contracts/external/actions/TreasuryAction.sol

```diff
@@ -0,0 +1,116 @@
+        WETH = _weth;
+        treasuryManagerContract = address(0);
+    }
+
+    function claimCOMP(address[] calldata ctokens) external override onlyManagerContract returns (uint256) {
+        COMPTROLLER.claimComp(address(this), ctokens);
+        uint256 bal = COMP.balanceOf(address(this));
+        COMP.transfer(treasuryManagerContract, bal);
+        return bal;
+    }
+
+    function transferReserveToTreasury(address[] calldata assets)
+        external
+        override
+        onlyManagerContract
+        returns (uint256[] memory)
+    {
+        uint256[] memory amountsTransferred = new uint256[](assets.length);
+
+        for (uint256 i; i < assets.length; i++) {
+            uint16 currencyId = tokenAddressToCurrencyId[assets[i]];
+            require(currencyId != 0, "Token not listed");
+
+            // prettier-ignore
+            (int256 reserve, /* */, /* */, /* */) = BalanceHandler.getBalanceStorage(Constants.RESERVE, currencyId);
+            Token memory token = TokenHandler.getAssetToken(currencyId);
+
+            uint256 totalReserve = reserve.toUint();
+            uint256 totalBalance = IERC20(token.tokenAddress).balanceOf(address(this));
+            uint256 buffer = reserveBuffer[token.tokenAddress];
+
+            // Reserve requirement not defined
+            if (buffer == 0) 
+                continue;
+            
+            uint256 requiredReserve = totalBalance.mul(buffer).div(uint256(Constants.INTERNAL_TOKEN_PRECISION));
+            
+            if (totalReserve > requiredReserve) {
+                uint256 redeemAmount = totalReserve.sub(requiredReserve);
+                Token memory underlying = TokenHandler.getUnderlyingToken(currencyId);
+                amountsTransferred[i] = _redeemCToken(token.tokenAddress, underlying.tokenAddress, redeemAmount);
+
+                // _redeemCToken wraps ETH into WETH
+                address underlyingAddress = underlying.tokenAddress == address(0) ? WETH : underlying.tokenAddress;
+                IERC20(underlyingAddress).safeTransfer(treasuryManagerContract, amountsTransferred[i]);
+            }
+        }
+
+        return amountsTransferred;
+    }
+
+    function _redeemCToken(address asset, address underlying, uint256 amount) internal returns (uint256) {
+        CErc20Interface(asset).redeem(amount);
+        uint256 redeemed;
+
+        if (underlying == address(0))
+        {
+            redeemed = address(this).balance;
+            WETH9(WETH).deposit{value: redeemed}();
+        }
+        else
+        {
+            redeemed = IERC20(underlying).balanceOf(address(this));
+        }
+
+        return redeemed;
+    }
+
+    function setTreasuryManager(address manager) external override onlyOwner {
+        treasuryManagerContract = manager;
+    }
+
+    function setReserveBuffer(address asset, uint256 amount) external override onlyOwner {
+        require(amount <= uint256(Constants.INTERNAL_TOKEN_PRECISION), "amount too large");
```

Comment:
Let's have this keyed by currency id instead.

Reply:
fixed